\clearpage
\section{Tool availability and implementation}
\label{sec.tool}
\smashpp is implemented in \cpp language and is available at~\cite{web-smashpp}. This tool is able to find and visualize rearrangements in sequences, \fasta and \fastq files; although, it is highly recommended to use sequences as input. In the following sections, we describe installing and running the \smashpp tool.

\subsection{Install}
To install \smashpp on various operating systems, follow the instructions below. Note that the precompiled executables are available for 64 bit operating systems in the ``bin/'' directory.

% \paragraph{Conda}
% % todo

\subsubsection*{Linux}
\begin{itemize}
  \item Install ``git'' and ``cmake'':
\begin{code}[style=bash]
sudo apt update
sudo apt install git cmake
\end{code}
\item Clone \smashpp and install it:
\begin{code}[style=bash]
git clone https://github.com/smortezah/smashpp.git
cd smashpp
./install.sh
\end{code}
\end{itemize}

\subsubsection*{macOS}
\begin{itemize}
    \item Install ``Homebrew'', ``git'' and ``cmake'':
\begin{code}[style=bash]
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
brew install git cmake
\end{code}
\item Clone \smashpp and install it:
\begin{code}[style=bash]
git clone https://github.com/smortezah/smashpp.git
cd smashpp
./install.sh
\end{code}
\end{itemize}

\subsubsection*{Windows}
\begin{itemize}
  \item Download and install ``CMake'', e.g., from https://github.com/Kitware/CMake/releases/\linebreak download/v3.14.4/cmake-3.14.4-win64-x64.msi. Make sure to add it to the system PATH. For example, if CMake is installed in ``C:\textbackslash Program Files'', add ``C:\textbackslash Program Files\textbackslash CMake\textbackslash bin'' to the system PATH.
  \item Download and install ``mingw-w64'', e.g., from https://sourceforge.net/projects/mingw-w64/\linebreak files/latest/download. Make sure to add it to the system PATH. For example, if it is installed in ``C:\textbackslash mingw-w64'', add ``C:\textbackslash mingw-w64\textbackslash mingw64\textbackslash bin'' to the system PATH.
  \item Download and install ``git'', from https://git-scm.com/download/win.
  \item Clone \smashpp and install it:
\begin{code}[style=bash]
git clone https://github.com/smortezah/smashpp.git
cd smashpp
.\install.bat
\end{code}
\end{itemize}

\subsection{Run \smashpp}
Various options are provided with the interface of the proposed tool, which are described in Table~\ref{tab.options}.

\begin{longtable}{lp{3cm}l}
  \caption{Options.}
  \label{tab.options} \\
  \toprule
  Flag & Input & Description \\
  \midrule
  \mono{-r} $^*$ & Seq/FASTA/FASTQ file & reference file. It is highly recommended to have short name. \\
  \mono{-t} $^*$ & Seq/FASTA/FASTQ file  & target file. It is highly recommended to have short name. \\
  \mono{-l} & integer in $[0, 5]$ & level of compression (default: 0). \\
  \mono{-m} & integer in $[1, 2^{32}-1]$ & minimum segment size (default: 50). Only those regions in the reference file that are bigger than this value would be able to be considered for compression. \\
  \mono{-e} & float in $[0.0, 100.0]$ & entropy of `N's (default: 2.0). In implementation of the reference-based compression, we have replaced `N' bases in the references and the targets with `A's and `T's, respectively. On reference-free compression, they are replaced with `A's, in both references and targets. If a user tends to replace `N' bases in a sequence with a normal distribution of `A', `C', `G' and `T's, he/she can employ GOOSE toolkit~\cite{web-goose}. \\
  \mono{-n} & integer in $[1, 8]$ & number of threads (default: 4). Creating multiple finite-context models and substitutional-tolerant Markov models can be done in a multi-threaded fashion. \\
  \mono{-f} & integer in $[1, 2^{32}-1]$ & filter size (default: 256). In the process of finding similar regions in the reference and the target sequences, the information content that would be obtained by compression needs to be filtered. \\
  \mono{-ft} & integer/string in \{0$|$rectangular, 1$|$hamming, 2$|$hann, 3$|$blackman, 4$|$triangular, 5$|$welch, 6$|$sine, 7$|$nuttall\} & filter type/windowing function (default: hann). Besides Hann window that is used by default to smooth the information content (profile), we have implemented several other windowing functions: Blackman~\cite{blackman1959particular}, Hamming~\cite{tukey1949measuring}, Nuttall~\cite{nuttall1981some}, rectangular~\cite{oppenheim1999discrete}, sine~\cite{harris1978use}, triangular~\cite{bartlett1950periodogram} and Welch~\cite{welch1967use} windows. These functions are given by Equation~\ref{eq.filter} and are plotted in Fig.~\ref{fig.filters}.
  \\
  \mono{-fs} & char/string in \{S$|$small, M$|$medium, L$|$large\} & filter scale (default: large) \\
  \mono{-d} & integer in $[1, 2^{64}-1]$ & sampling steps (default: 1). \\
  \mono{-th} & float in $[0.0, 20.0]$ & threshold (default: 1.5).\\
  \mono{-rb} & integer in $[-2^{15}, 2^{15}-1]$ & reference beginning guard (default: 0). \\
  \mono{-re} & integer in $[-2^{15}, 2^{15}-1]$ & reference ending guard (default: 0). \\
  \mono{-tb} & integer in $[-2^{15}, 2^{15}-1]$ & target beginning guard (default: 0). \\
  \mono{-te}       & integer in $[-2^{15}, 2^{15}-1]$                                                                                             & target ending guard (default: 0)                                                                                                                                                                                                                          \\
  \mono{-dp}       & --- $^\dagger$                                                                                                               & deep compression                                                                                                                                                                                                                                          \\
  \mono{-nr}       & ---                                                                                                                          & do NOT compute self complexity                                                                                                                                                                                                                            \\
  \mono{-sb}       & ---                                                                                                                          & save sequence (input: FASTA/FASTQ)                                                                                                                                                                                                                        \\
  \mono{-sp}       & ---                                                                                                                          & save profile (*.prf)                                                                                                                                                                                                                                      \\
  \mono{-sf}       & ---                                                                                                                          & save filtered file (*.fil)                                                                                                                                                                                                                                \\
  \mono{-sf}       & ---                                                                                                                          & save segmented files (*.s$_i$)                                                                                                                                                                                                                            \\
  \mono{-sa}       & ---                                                                                                                          & save profile, filetered and segmented files                                                                                                                                                                                                               \\
  \mono{-rm}       & $k$,[$w$,$d$,]ir,$a$,$g$/$t$,ir,$a$,$g$:...                                                                                  & parameters of reference models (default: $14,0,0.005,0.95$):
  $k$ (integer):  context size
  $w$ (integer):  width of sketch in $\log_2$ form,
  e.g., set 10 for $w=2^10=1024$
  $d$ (integer):  depth of sketch
  ir (integer in $\{0, 1, 2\}$): inverted repeat
  0: regular (not inverted)
  1: inverted, solely
  2: both regular and inverted
  $a$ (float $>0$):  estimator
  $g$ (float in $[0.0, 1.0)$):  forgetting factor
  $t$ (integer):  threshold (no. substitutions)
  \\
  \mono{-tm}       & $k$,[$w$,$d$,]ir,$a$,$g$/$t$,ir,$a$,$g$:...                                                                                  & parameters of target models (default: $14,0,0.005,0.95$)
  \\
  \mono{-ll}       & ---                                                                                                                          & save list of compression levels                                                                                                                                                                                                                           \\
  \mono{-h}        & ---                                                                                                                          & usage guide                                                                                                                                                                                                                                               \\
  \mono{-v}        & ---                                                                                                                          & more information (verbose)                                                                                                                                                                                                                                \\
  \mono{--version} & ---                                                                                                                          & show version                                                                                                                                                                                                                                              \\
  \bottomrule
  \vspace{-2mm}
  % $^*$ Mandatory, $^\dagger$ no input, $\times$ not used
\end{longtable}

\begin{align} \label{eq.filter}
  w[n] & = 1,
  \tag*{(rectangular)}                                                                                                                                        \\
  w[n] & = 1-\left|\tfrac {n-N/2}{L/2}\right|, \quad L=N,
  \tag*{(triangular/Bartlett)}                                                                                                                                \\
  w[n] & = 1-\left(\tfrac {n-N/2}{N/2}\right)^{2},
  \tag*{(Welch)}                                                                                                                                              \\
  w[n] & = \sin \left(\tfrac {\pi n}{N}\right),
  \tag*{(sine)}                                                                                                                                               \\
  w[n] & = 0.54348-0.45652\;\cos \left(\tfrac {2\pi n}{N}\right),
  \tag*{(Hamming)}                                                                                                                                            \\
  w[n] & = 0.42659-0.49656\;\cos \left(\tfrac {2\pi n}{N}\right)+0.07685\;\cos \left(\tfrac {4\pi n}{N}\right),
  \tag*{(Blackman)}                                                                                                                                           \\
  w[n] & = 0.35577-0.48740\;\cos \left(\tfrac {2\pi n}{N}\right)+0.14423\;\cos \left(\tfrac {4\pi n}{N}\right)-0.01260\;\cos \left(\tfrac {6\pi n}{N}\right),
  \tag*{(Nuttall)}                                                                                                                                            \\
\end{align}

A reference file and a target file are clearly mandatory to run \smashpp (without visualization). Running
\begin{code}[style=bash]
./smashpp
\end{code}
provides the following guide:
\begin{code}[style=bash]
SYNOPSIS
  ./smashpp [OPTIONS]  -r <REF-FILE>  -t <TAR-FILE>

OPTIONS
  Required:
  -r  <FILE>         = reference file (Seq/FASTA/FASTQ)
  -t  <FILE>         = target file    (Seq/FASTA/FASTQ)

  Optional:
  -l  <INT>          = level of compression: [0, 5]. Default -> 0
  -m  <INT>          = min segment size: [1, 4294967295]     -> 50
  -e  <FLOAT>        = entropy of 'N's: [0.0, 100.0]         -> 2.0
  -n  <INT>          = number of threads: [1, 8]             -> 4
  -f  <INT>          = filter size: [1, 4294967295]          -> 256
  -ft <INT/STRING>   = filter type (windowing function):     -> hann
                       {0|rectangular, 1|hamming, 2|hann,
                       3|blackman, 4|triangular, 5|welch,
                       6|sine, 7|nuttall}
  -fs [S][M][L]      = filter scale:                         -> L
                       {S|small, M|medium, L|large}
  -d  <INT>          = sampling steps                        -> 1
  -th <FLOAT>        = threshold: [0.0, 20.0]                -> 1.5
  -rb <INT>          = ref beginning guard: [-32768, 32767]  -> 0
  -re <INT>          = ref ending guard: [-32768, 32767]     -> 0
  -tb <INT>          = tar beginning guard: [-32768, 32767]  -> 0
  -te <INT>          = tar ending guard: [-32768, 32767]     -> 0
  -dp                = deep compression                      -> no
  -nr                = do NOT compute self complexity        -> no
  -sb                = save sequence (input: FASTA/FASTQ)    -> no
  -sp                = save profile (*.prf)                  -> no
  -sf                = save filtered file (*.fil)            -> no
  -ss                = save segmented files (*.s[i])         -> no
  -sa                = save profile, filetered and           -> no
                       segmented files
  -rm k,[w,d,]ir,a,g/t,ir,a,g:...
  -tm k,[w,d,]ir,a,g/t,ir,a,g:...
                     = parameters of models
                <INT>  k:  context size
                <INT>  w:  width of sketch in log2 form,
                           e.g., set 10 for w=2^10=1024
                <INT>  d:  depth of sketch
                <INT>  ir: inverted repeat: {0, 1, 2}
                           0: regular (not inverted)
                           1: inverted, solely
                           2: both regular and inverted
              <FLOAT>  a:  estimator
              <FLOAT>  g:  forgetting factor: [0.0, 1.0)
                <INT>  t:  threshold (no. substitutions)
  -ll                = list of compression levels
  -h                 = usage guide
  -v                 = more information
  --version          = show version
\end{code}

The arguments ``-r'' and ``-t'' are used to specify the reference and the target, respectively, which are highly recommended to have short names. 
Level of compression, that is an integer between 0 and 5, can be determined with ``-l''. 
By setting ``-m'' to an integer value, only those regions in the reference file that are bigger than that value would be able to be considered for compression.

In implementation of the reference-based compression, we have replaced `N' bases in the references and the targets with `A's and `T's, respectively. On reference-free compression, they are replaced with `A's, in both references and targets. If a user tends to replace `N' bases in a sequence with a normal distribution of `A', `C', `G' and `T's, he/she can employ GOOSE toolkit~\cite{web-goose}. Note that we have set by default the entropy of `N's to 2.0, however, it can be changed to a value of interest using ``-e'' option.

Creating multiple finite-context models and substitutional-tolerant Markov models can be done in a multi-threaded fashion by setting ``-n'' to an integer. 

In the process of finding similar regions in the reference and the target sequences, the information content that would be obtained by compression needs to be filtered. Size of the window and type of the windowing function can be set by ``-f'' and ``-ft'' options, respectively.
Besides Hann window that is used by default to smooth the information content (profile), we have implemented several other windowing functions, including Blackman~\cite{blackman1959particular}, Hamming~\cite{tukey1949measuring}, Nuttall~\cite{nuttall1981some}, rectangular~\cite{oppenheim1999discrete}, sine~\cite{harris1978use}, triangular~\cite{bartlett1950periodogram} and Welch~\cite{welch1967use} windows. These functions are given by
\begin{align}
  w[n] & = 1,
  \tag*{(rectangular)}                                                                                                                                        \\
  w[n] & = 1-\left|\tfrac {n-N/2}{L/2}\right|, \quad L=N,
  \tag*{(triangular/Bartlett)}                                                                                                                                \\
  w[n] & = 1-\left(\tfrac {n-N/2}{N/2}\right)^{2},
  \tag*{(Welch)}                                                                                                                                              \\
  w[n] & = \sin \left(\tfrac {\pi n}{N}\right),
  \tag*{(sine)}                                                                                                                                               \\
  w[n] & = 0.54348-0.45652\;\cos \left(\tfrac {2\pi n}{N}\right),
  \tag*{(Hamming)}                                                                                                                                            \\
  w[n] & = 0.42659-0.49656\;\cos \left(\tfrac {2\pi n}{N}\right)+0.07685\;\cos \left(\tfrac {4\pi n}{N}\right),
  \tag*{(Blackman)}                                                                                                                                           \\
  w[n] & = 0.35577-0.48740\;\cos \left(\tfrac {2\pi n}{N}\right)+0.14423\;\cos \left(\tfrac {4\pi n}{N}\right)-0.01260\;\cos \left(\tfrac {6\pi n}{N}\right),
  \tag*{(Nuttall)}                                                                                                                                            \\
\end{align}
and are plotted in Fig.~\ref{fig.filters}.
Scale of the filter can be set as S (small), M (medium) or L (large), using ``-fs''.
Also, instead of considering the whole information content, the user is able to make samples of it by steps of which size can be determined by ``-d''.
\begin{figure}[!h]
  \centering
  \includegraphics[width=\linewidth]{filters.pdf}
  \caption{Various windowing functions implemented and embedded in \smashpp.}
  \label{fig.filters}
\end{figure}

For the purpose of segmenting the filtered information content, the average entropy of reference-based compression is used by default as the threshold, but the threshold can be altered by ``-th'' option.

\smashpp is capable of finding even very small similar regions in two sequences. However, we have found experimentally that when it is running in a very sensitive mode, there might be some cases in which the size of similar regions in the reference and the target are not balanced. These cases can be handled by ``-rb'', ``-re'', ``-tb'' and ``-te'' options, that can resize the beginning and ending guards of reference and target regions, respectively.
For example, if ``-tb 10'' is used, the first 10 bases in each target region will be ignored, which results in smaller regions.
Note that when the guard sizes of target regions are increased, the models built from these regions would be slightly different than the original models; consequently, the sizes of reference regions that are detected as being similar to the ones from the target would be modified, as well. Therefore, changing the guard sizes of target regions will affect the sizes of reference regions. 
In the case of activating deep compression, by ``-dp'', changing the guard sizes of reference regions would affect the sizes of target regions, as well.

The ``deep compression'' means that similar regions in target and reference sequences are found in three phases instead of two: 1) the model of the reference is built and the target is compressed based upon that model, 2) the model of each detected region is built and the whole reference is compressed based on these models and 3) the model of each detected reference region is built and the corresponding target regions will be compressed based on that model.

Triggering ``-nr'' makes the tool not to perform the reference-free compression (self-complexity computation).

\smashpp accepts \fasta and \fastq files as input, in addition to sequences. In these cases, the input files are first converted to sequences and then processed further. It is possible to save these sequences by ``-sb'' option.
When the information profile is obtained, \smashpp smoothens then removes it by default. However, it can be preserved by ``-sp'' option.
The same thing happens to the smoothed file, i.e., it is segmented then removed. But, the user can use ``-sf'' to save the filtered file. 
Also, the segmented files can be saved using ``-ss''. 
The user can save all the information profile (content), filtered and segmented files, by triggering ``-sa'' option.

For the purpose of compression, it is recommended to use ``-l'' option, since it configures the models automatically. However, using ``-rm'' and ``-tm'', the user would be able to manually configure the reference model, for reference-based compression, and the target model, for reference-free compression, respectively. Parameters of the models are described in detail in section~\ref{sec.methods}.
Note that using ``-ll'' option, the list of parameters that would be chosen for each model automatically, will be shown.

By running \smashpp, positions of the similar regions in reference and target sequences, and also complexity of the regions is saved in a ``.pos'' file. This tab-separated file has a header including:
\begin{enumerate}
  \item The string ``\#SMASHPP'' as a specifier for the \smashpp tool,
  \item The name of reference sequence,
  \item The size of reference sequence,
  \item The name of target sequence,
  \item The size of target sequence,
\end{enumerate}
and a body including:
\begin{enumerate}
  \item Initial position of the reference sequence,
  \item Final position of the reference sequence,
  \item Average entropy of compressing the associated target block considering this reference block as the reference,
  \item Complexity (average entropy) of the detected reference block, calculated by reference-free compression,
  \item Initial position of the target sequence,
  \item Final position of the target sequence,
  \item Average entropy of compressing the associated reference block considering this target block as the reference,
  \item Complexity of the detected target block, calculated by reference-free compression.
\end{enumerate}
As an example, the header of the following ``.pos'' file shows that the reference ``Ref'' and the target ``Tar'' are 5,000,000 bases long. The body shows that there is a block in the Ref, from the position 1250013 up to 2499988, which is similar to a block in the Tar, from the position 1250017 to 2499970. Average entropy of compressing the Tar block, using the Ref block as reference, is 0.238067. This number is 0.26698 when the Ref block is compressed based on the model of the Tar block. Also, complexities of the Ref and the Tar blocks are 1.97158 and 1.98884, respectively.
\begin{code}[style=bash]
#SMASHPP Ref      5000000  Tar      5000000
1250013	 2499988  0.26698  1.97158  1250017  2499970  0.238067 1.98884
\end{code}


\subsection{Run \smashpp visualizer}
Running \smashpp (without visualization), positions of the similar regions in reference and target sequences, and also complexity of the regions is saved in a *.pos file, that can be visualized by
\begin{code}[style=bash]
./smashpp -viz
\end{code}
which gives
\begin{code}[style=bash]
SYNOPSIS
  ./smashpp -viz [OPTIONS]  -o <SVG-FILE>  <POS-FILE>

OPTIONS
  Required:
  <POS-FILE>         = position file, generated by
                       Smash++ tool (*.pos)

  Optional:
  -o  <SVG-FILE>     = output image name (*.svg)             -> map.svg
  -rn <STRING>       = reference name shown on output. If it
                       has space, use double quotes, e.g.
                       "Seq label". Default: name in header
                       of position file
  -tn <STRING>       = target name shown on output
  -l  <INT>          = type of the link between maps: [1, 6] -> 1
  -c  <INT>          = color mode: [0, 1]                    -> 0
  -p  <FLOAT>        = opacity: [0.0, 1.0]                   -> 0.9
  -w  <INT>          = width of the sequence: [15, 100]      -> 16
  -s  <INT>          = space between sequences: [15, 200]    -> 62
  -f  <INT>          = multiplication factor for             -> 43
                       color ID: [1, 255]
  -b  <INT>          = beginning of color ID: [0, 255]       -> 0
  -rt <INT>          = reference tick: [1, 4294967295]
  -tt <INT>          = target tick: [1, 4294967295]
  -th [0][1]         = tick human readable: 0=false, 1=true  -> 1
  -m  <INT>          = minimum block size: [1, 4294967295]   -> 1
  -vv                = vertical view                         -> no
  -nn                = do NOT show normalized relative       -> no
                       compression (NRC)
  -nr                = do NOT show self complexity           -> no
  -ni                = do NOT show inverse maps              -> no
  -ng                = do NOT show regular maps              -> no
  -h                 = usage guide
  -v                 = more information
  --version          = show version
\end{code}

The output of \smashpp visualizer is an ``SVG'' file whose name is determined by ``-o'' option. By default, it is named ``map.svg''. 
Names of the reference and the target, which are going to be printed on the output image, can be altered by ``-rn'' and ``-tn'', respectively. They are by default the names written in the positions file. 

Options ``-l'', ``-c'', ``-p'', ``-w'', ``-s'', ``-f'' and ``-b'' can be used to change the appearance of the image.

Assigning integers to ``-rt'' and ``-tt'' options will change the tick sizes of the reference and the target, respectively. \smashpp prints the sizes on axes in human readable format, e.g., 1K, 2M, etc. However, it can be triggered by ``-th'' option. Note that, here, 1K is equivalent to 1000 and not 1024, and so on. 

By setting ``-m'' to an integer value, only the regions that are bigger than that value will be illustrated.
To have a vertical view of the image, instead of the default horizontal view, one can use ``-vv'' trigger.

\smashpp performs reference-based and reference-free compressions to calculate the normalized relative compression (NRC) and redundancy (self complexity), respectively. If the user is not interested in showing them, he/she can turn them off by ``-nn'' and ``-nr'' triggers. In addition, \smashpp considers by default both regular and reverse complement maps in its calculations. Triggering ``-ni'' and ``-ng'' will stop showing inverted and regular maps, respectively.

\subsection{Example}
This section guides step-by-step employing \smashpp to find and visualize rearrangements in a sample genomic data. Note that the commands can be run on Linux and macOS, however, they are similar in Windows.

\subsubsection*{Install \smashpp and provide the required files}
First, install \smashpp:
\begin{code}[style=bash]
git clone https://github.com/smortezah/smashpp.git
cd smashpp
./install.sh
\end{code}
Then, copy \mono{smashpp} executable file into \mono{example/} directory and go to that directory:
\begin{code}[style=bash]
cp smashpp example/
cd example/
\end{code}
There is a 1000 byte reference sequence, named \mono{refs}, and a 1000 byte target sequence, named \mono{tars}, in this directory. Running
\begin{code}[style=bash]
./smashpp -r refs -t tars -f 45 -l 3
./smashpp -viz -p 1 -s 50 -w 15 refs.tars.pos
\end{code}
results in Fig.~\ref{fig.example}, which has been saved as ``map.svg''.

\begin{figure}[!h]
  \includegraphics[width=.95\linewidth]{example.pdf}
  \caption{An example of running \smashpp on two 1000 base sequences. Two similar regions in regular mode and two similar ones in inverted mode are detected.}
  \label{fig.example}
\end{figure}